demos/common/utilities/jws_using_mbedtls.c

```diff
@@ -0,0 +1,634 @@
+
+    LogInfo( ( "--- Checking if SHA256 of header+payload matches decrypted SHA256 ---\n" ) );
+
+    /* TODO: remove this once we have a valid PKCS7 parser. */
+    int doTheyMatch = memcmp( pucBuffer + jwsPKCS7_PAYLOAD_OFFSET, pucShaBuffer, jwsSHA256_SIZE );
+
+    if( doTheyMatch == 0 )
+    {
+        LogInfo( ( "SHA of JWK matches\n" ) );
+        xResult = 0;
+    }
+    else
+    {
+        LogError( ( "SHA of JWK does NOT match\n" ) );
+        xResult = 1;
+    }
+
+    return xResult;
+}
+
+
+uint32_t JWS_ManifestAuthenticate( const char * pucManifest,
+                                   uint32_t ulManifestLength,
+                                   char * pucJWS,
+                                   uint32_t ulJWSLength )
+{
+    uint32_t ulVerificationResult;
+
+    int lMbedResult;
+    unsigned char * pucBase64EncodedHeader;
+    unsigned char * pucBase64EncodedPayload;
+    unsigned char * pucBase64EncodedSignature;
+    uint32_t ulBase64EncodedHeaderLength;
+    uint32_t ulBase64EncodedPayloadLength;
+    uint32_t ulBase64SignatureLength;
+    AzureIoTJSONReader_t xJSONReader;
+
+    LogInfo( ( "---------------------Begin Signature Validation --------------------\n\n" ) );
+
+    /*------------------- Parse and Decode the Manifest Sig ------------------------*/
+
+    AzureIoTResult_t xResult = prvSplitJWS( ( unsigned char * ) pucJWS, ulJWSLength,
+                                            &pucBase64EncodedHeader, &ulBase64EncodedHeaderLength,
+                                            &pucBase64EncodedPayload, &ulBase64EncodedPayloadLength,
+                                            &pucBase64EncodedSignature, &ulBase64SignatureLength );
+    prvSwapToUrlEncodingChars( pucBase64EncodedSignature, ulBase64SignatureLength );
+
+    /* Note that we do not use mbedTLS to base64 decode values since we need the ability to assume padding characters. */
+    /* mbedTLS will stop the decoding short and we would then need to hack in the remaining characters. */
+    LogInfo( ( "---JWS Base64 Decode Header---\n" ) );
+    int32_t outJWSHeaderLength;
+    az_span xJWSBase64EncodedHeaderSpan = az_span_create( pucBase64EncodedHeader, ulBase64EncodedHeaderLength );
+    az_span xJWSHeaderSpan = az_span_create( ucJWSHeader, sizeof( ucJWSHeader ) );
+    az_result xCoreResult = az_base64_decode( xJWSHeaderSpan, xJWSBase64EncodedHeaderSpan, &outJWSHeaderLength );
+    /* TODO: here and elsewhere, remove verbose logs */
+    /* LogInfo(( "az_base64_decode return: 0x%x\n", xCoreResult)); */
+    /* LogInfo(( "\tOut Decoded Size: %i\n", outJWSHeaderLength )); */
+    /* LogInfo(( "\t%.*s\n\n", ( int ) outJWSHeaderLength, ( char * ) ucJWSHeader )); */
+
+    LogInfo( ( "---JWS Base64 Decode Payload---\n" ) );
+    int32_t outJWSPayloadLength;
+    az_span xJWSBase64EncodedPayloadSpan = az_span_create( pucBase64EncodedPayload, ulBase64EncodedPayloadLength );
+    az_span xJWSPayloadSpan = az_span_create( ucJWSPayload, sizeof( ucJWSPayload ) );
```

Comment:
First you should checkout comments I have about whether to use statics for stuf8f like `ucJWSPayload` at top of this file.

That said, we do this parse quite early in the process and only start using the buffer like 200 lines later, like line 578.  Could we potentially only actually do unencode when we need to?  The advantage of this is that we could have a single scratch buffer, so we only do the base64_decode and then immediately do JSON parsing on it.  And at other places we need the scratch, we do the calculation in that and then we're done.  So we end up needing less memory than having each buffer allocated on its own separately.

Reply:
Yea it's a tradeoff here. Doing it like this makes the flow easier to follow (you're just moving down the layers of the JWS). Breaking it up and doing only what's necessary means kicking up and down the "stack" of the JWS. I can try and rework to reduce memory usage.

Reply:
In general I'm all for flow in code (and I try to code that way myself if not always successfully which is why you guys keep me honest:)).  But I think/hope here that once this long function gets broken up into smaller helpers, then the flow will me more natural to read AND we can get memory savings at same time.  Though given it's 100's of bytes we can save, I think saving that has to be guiding principle if there's no clever way to flow+save mem at same time.

Reply:
I reduced the usage in the latest iteration by moving some of the parsing (`3875->3358` for `~500` bytes). It's possible that more could be done by parsing things out and memcpy'ing relevant values to other parts of the scratch space. But that would probably require a greater amount of fine-tooth combing. I'm willing to go down that rabbit hole but it might be another iteration.